### DISCUSSION TITLE
Cutting image into tiles

### DISCUSSION THREAD

**Discussion by aradep** [libvips/libvips#4268] [Q&A]
I'm trying to write a bash script that uses libvips to cut large (16k) texturemap images into 256x256 tiles (16bit grayscale .png) named as "prefix_x_y_suffix" with prefix and suffix being user-defined variables.

I found some info on the dzsave function and it completes in 2.2 seconds what takes ImageMagick 30 seconds, the only problem is the lack of prefix and the forced directory structure. Moving/renaming the files afterwards makes the script take 20-30s longer. Is there a way to make dzsave do what I want, or is there another function that would work for this? 

Thanks

**Comment by jcupitt** [comment]
Hello @aradep,

I would use dzsave and then rename the tiles. It should be quick, I'd think, let me have a quick go.

**Comment by jcupitt** [comment]
I tried here with a 16k x 16k pixel 16-bit mono vips file:

```
$ vipsheader ~/pics/16k-g16.v
/home/john/pics/16k-g16.v: 16384x16384 ushort, 1 band, grey16
```

I see:

```
$ time time vips dzsave ~/pics/16k-g16.v x \
  --overlap 0 --tile-size 256 --depth one --layout dz --suffix .png

real	0m9.232s
user	1m14.621s
sys	0m3.115s
```

I'm surprised it ran so quickly for you -- did you set the suffix? It defaults to JPG tiles, which are maybe 20x faster. 

I tried a rename script:

```bash
#!/bin/bash

set -e

infile=$1
outdir=$2
prefix=$3
suffix=$4

width=$(vipsheader -f width $1)
height=$(vipsheader -f height $1)
max_x=$(($width / 256 + ($width % 256 > 0) - 1))
max_y=$(($height / 256 + ($height % 256 > 0) - 1))

echo infile = $infile
echo outdir = $outdir
echo prefix = $prefix
echo suffix = $suffix
echo max_x = $max_x
echo max_y = $max_y

echo generating tiles ...
rm -rf ${outdir}_files
vips dzsave $infile $outdir \
  --overlap 0 --tile-size 256 --depth one --layout dz --suffix .png

echo renaming ...
for x in $(seq 0 $max_x); do
  for y in $(seq 0 $max_y); do
    mv ${outdir}_files/0/${x}_${y}.png \
      ${outdir}_files/0/${prefix}_${x}_${y}_${suffix}.png
  done
done
```

I see:

```
$ time ./tile.sh ~/pics/16k-g16.v x fred jim 
infile = /home/john/pics/16k-g16.v
outdir = x
prefix = fred
suffix = jim
max_x = 63
max_y = 63
generating tiles ...
renaming ...

real	0m15.210s
user	1m16.135s
sys	0m8.279s
```

So about 9s for dzsave and 6s for the rename. It renames tiles to:

```
$ vipsheader x_files/0/fred_0_0_jim.png 
x_files/0/fred_0_0_jim.png: 256x256 ushort, 1 band, grey16, pngload
```

Is that what you wanted? I may have misunderstood.

**Comment by aradep** [comment]
Hi, yes this is what I wanted.  I tried your renaming script but for me the rename takes 1m30s. Moving the files without renaming takes 2s. Converting the .png I was using into a vips file had no improvement. (This is Windows 10 and the files are on a sata ssd non-boot drive.)

>I'm surprised it ran so quickly for you

Sorry had my numbers mixed up. 16k grayscale is about 4.5s with vips and 17s with imagemagick.

**Reply by jcupitt** [reply]
Ah OK. Windows is generally slower at file operations. You could try in WSL, it should be a lot quicker.
